gtsfm/frontend/detector_descriptor/d2net.py

```diff
@@ -0,0 +1,117 @@
+
+    def __init__(self, max_keypoints: int = 5000, model_path: Path = MODEL_PATH, use_cuda: bool = True) -> None:
+        """ """
+        super().__init__()
+        self.max_keypoints = max_keypoints
+        self.model_path = model_path
+        self.use_cuda = use_cuda and torch.cuda.is_available()
+        self.device = torch.device("cuda" if self.use_cuda else "cpu")
+
+    def detect_and_describe(self, image: Image) -> Tuple[Keypoints, np.ndarray]:
+        """Extract keypoints and their corresponding descriptors.
+
+        Adapted from:
+        https://github.com/mihaidusmanu/d2-net/blob/master/extract_features.py
+
+        Args:
+            image: the input image.
+
+        Returns:
+            Detected keypoints, with length N <= max_keypoints.
+            Corr. descriptors, of shape (N, D) where D is the dimension of each descriptor.
+        """
+        model = D2Net(model_file=self.model_path, use_relu=USE_RELU, use_cuda=self.use_cuda)
+        model.eval()
+
+        # Resize image, and obtain re-scaling factors to postprocess keypoint coordinates.
+        resized_image, fact_i, fact_j = resize_image(image.value_array)
+        input_image = preprocess_image(resized_image, preprocessing=PREPROCESSING_METHOD)
+
+        if USE_MULTISCALE:
+            scales = PYRAMID_SCALES
+        else:
+            scales = [1]
+
+        with torch.no_grad():
+            keypoints, scores, descriptors = d2net_pyramid.process_multiscale(
+                image=torch.tensor(input_image[np.newaxis, :, :, :].astype(np.float32), device=self.device),
+                model=model,
+                scales=scales,
+            )
+
+        # Choose the top K keypoints and descriptors.
+        ordered_idxs = np.argsort(-scores)[: self.max_keypoints]
+        keypoints = keypoints[ordered_idxs, :]
+        descriptors = descriptors[ordered_idxs, :]
+
+        # Rescale keypoint coordinates from resized image scale, back to provided input image resolution.
+        keypoints[:, 0] *= fact_i
+        keypoints[:, 1] *= fact_j
+
+        # Convert (y,x) tuples that represented (i, j) indices of image matrix, into (u, v) coordinates.
+        keypoints = keypoints[:, [1, 0]]
+        return Keypoints(coordinates=keypoints), descriptors
+
+
+def resize_image(image: np.ndarray) -> Tuple[np.ndarray, float, float]:
+    """Resize image to limit to a maximum edge length and maximum summed length of both edges.
+
+    Args:
+        image: array of shape (H,W,3) representing an RGB image.
+
+    Returns:
+        resized_image: resized image.
+        fact_i: vertical re-scaling factor for keypoint x-coordinates (inverse of downsampling factor).
+        fact_j: horizontal re-scaling factor for keypoint y-coordinates.
+    """
+    if len(image.shape) == 2:
+        # If grayscale, repeat grayscale channel 3 times.
+        image = image[:, :, np.newaxis]
+        image = np.repeat(image, 3, -1)
+
+    resized_image = image
```

Comment:
is that a deepcopy? does changing resized_image change image?

Reply:
why not use image as is? scipy must do the deepcopy

Reply:
Oh that's to show that the default behavior is to treat the original image as the resized image. Then, if needed, the resized image is updated.

Reply:
We already have image resize function in the utils, right? Why not use that?

Reply:
Good question. So all of our resize APIs work around providing fixed dimensions (the logic here is about providing a fractional scale), or limiting the short side of the image. I'd rather just leave it closer to the author's impl and avoid re-implementing all that logic.